WalletWasabi/Blockchain/TransactionBuilding/SmartCoinSelector.cs

```diff
@@ -17,18 +17,34 @@ public SmartCoinSelector(List<SmartCoin> unspentCoins)
 	}
 
 	private List<SmartCoin> UnspentCoins { get; }
+	private int IterationCount { get; set; }
 
-	/// <param name="unused">Unused parameter, make it an empty list.</param>
-	public IEnumerable<ICoin> Select(IEnumerable<ICoin> unused, IMoney target)
+	/// <param name="suggestion">We use this to detect if NBitcoin tries to suggest something different and indicate the error.</param>
+	public IEnumerable<ICoin> Select(IEnumerable<ICoin> suggestion, IMoney target)
 	{
-		var targetMoney = target as Money;
+		var targetMoney = (Money)target;
 
 		long available = UnspentCoins.Sum(x => x.Amount);
 		if (available < targetMoney)
 		{
 			throw new InsufficientBalanceException(targetMoney, available);
 		}
 
+		// This is a hacky safety check to make sure the first few iterations can go through.
+		if (IterationCount > 10)
+		{
+			var suggestedSum = suggestion.Sum(c => (Money)c.Amount);
+			if (suggestedSum < (Money)target)
+			{
+				throw new InsufficientBalanceException(targetMoney, suggestedSum);
+			}
+		}
+
+		if (IterationCount > 500)
```

Comment:
Shouldn't we check this before `if (IterationCount > 10)`?

Reply:
ACK